This notebook, is to help the novice python learners to pick up python faster by showing them how to explore a dataset which aims to find patterns and relationships in data by using Python & its libraries. As this is the first notebook, we will cover the basic commands of Exploratory Data Analysis (EDA) which includes cleaning, munging, combining, reshaping, slicing and dicing, and transforming data for analysis purpose.


In order to improve any model's accuracy, we need to spend significant time on exploring and analyzing the Data. EDA is a critical & first step in analyzing the data and we do this for below reasons :

• Finding patterns in Data 
• Determining relationships in Data
• Checking of assumptions
• Preliminary selection of appropriate models
• Detection of mistakes


Data Exploration/Analysis is basically finding out more about the data we have.

And for Data Analysis, Python has gathered a lot of interest recently as a choice of language. Python’s improved library support (primarily pandas) has made it a strong alternative for data manipulation tasks. Combined with Python’s strength in general purpose programming, it is an excellent choice as a single language for building data-centric applications. 

We will be using following set of libraries :

Pandas: Extensively used for Structured data operations & manipulations and also for data munging & preparation

Numpy : Numpy stands for Numerical Python. This library contains basic linear algebra functions, Fourier Transforms, Advanced           Random Number Capabilities

Matplotlib : Python based plotting library offers matplotlib with complete 2D & limited 3D graphic support

SciPy : Collection of packages addressing a number of different standard problem domains in scientific computing. (Not used in this notebook but we will learn this later)

In [1]:
# Importing Standard Library functions & Displaying Pandas Version
import pandas as pd
import numpy as np
import sys


from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
#Here ast_note_interactivity kernel option make us see the value of multiple statements at once


print('Python version ' + sys.version)
print('Pandas version ' + pd.__version__)
print('NumPy  version ' + np.__version__)

Python version 2.7.12 |Anaconda 4.2.0 (64-bit)| (default, Jun 29 2016, 11:07:13) [MSC v.1500 64 bit (AMD64)]
Pandas version 0.19.2
NumPy  version 1.12.0


Just to begin with, we will analyse Chipotle Dataset for getting and knowing the Data which can help us drive valuable decisions for their business. You can download this Dataset from 
https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv


For importing data we use pd.read_csv(filename) function. 
Here pd refers to pandas as imported above`;`
     df refers to pandas dataframe object

In [5]:
# Read the chipotle dataset into pandas dataframe and see few sample records. This is in TSV(Tab Separated File) Form.


#location = r'C:\Users\SONY\Desktop\PANDAS\Python Pandas NumPy Jan 31\chipotle.tsv'
location = r'/Users/aurobindosarkar/Downloads/Nita/chipotle.tsv'
df = pd.read_csv(location, sep='\t')
df.head(5)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

We can see few sample records & this helps to understand the following :

• We are able to read the File
• We have used correct function to read the file
• Getting familiar with what values we might see. For e.g. item_price column is having `$` along with float values. So in case if we want to aggregate this column then we will have to get rid of the `$` sign.

--------------------------------------------------------------------------------------------------------------------------------



For inspecting Data we have below functions :

df.head(n)    -- First n rows of DataFrame

df.tail(n)    -- Last n rows of DataFrame

df.shape      -- Number of rows and columns

df.info()     -- Index, Datatype and Memory information

df.dtypes     -- Data Type of each Column

df.describe() -- Summary statistics for numerical columns

df.columns    -- Prints the name of all columns

df.ColumnName.value_counts(dropna=False) -- View unique values and counts
                                            dropna=False indicates Don’t include counts of NaN
                                            
df.apply(pd.Series.value_counts) -- Unique values and counts for all columns
 

Now Lets INSPECT Data by using above functions :

In [6]:
#Its important to understand the volume of data i.e. the number of rows & columns present in the given dataset which is given by shape function

df.shape
print('Number of Rows:    ' + str(df.shape[0]))
print('Number of Columns: ' + str(df.shape[1]))

(4622, 5)

Number of Rows:    4622
Number of Columns: 5


As you can see, everything has been read in properly – we have 4622 rows and 5 columns. Also this attribute enables us to access the metadata. And help us to decide if we can work on entire dataset or we need to do sampling for EDA in case the dataset is really very huge.

Sampling refers to the technique used to select, manipulate and analyze a representative subset of data in order to identify patterns and trends in the larger data set being examined. We will study this in detail in the upcoming notebook series.

In [7]:
print('Index, Datatype and Memory information : \n')
df.info()

# Memory usage of each column can be found by calling the below method.
# It returns a Series with an index represented by column names and memory usage of each column shown in bytes 
# And uses that as a launching point to improve the understanding of how each data types are stored in memory.
df.memory_usage()

Index, Datatype and Memory information : 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
order_id              4622 non-null int64
quantity              4622 non-null int64
item_name             4622 non-null object
choice_description    3376 non-null object
item_price            4622 non-null object
dtypes: int64(2), object(3)
memory usage: 180.6+ KB


This information is vital for our EDA as :

Indexing enables us to store and manipulate data with an arbitrary number of dimensions in lower dimensional data structures like Series (1d) and DataFrame (2d). Indexing also helps in faster identification of rows. 


Memory Usage tells us about an approximate amount of RAM used to hold the DataFrame which can even help to write more efficient code. 
Also this helps us to avoid accidental copies of data as it is one of the major cause of slowness.

When dealing with very large dataset /big data we need to keep a manageable in-memory data size and makes in-memory calculations fast & even multiple processing for cpu-intensive operations so understanding memory usage is critically important.
And for this we have few Python libraries like H5py, PyTables & mmap for memory management. (Not a part of this notebook)

In [8]:
df.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

Knowing these data types is highly important as they help us in knowing the numeric & non numeric fields. In case if there are any objects present, we need to convert them into desired data types. This would help us to use the data for any aggregate/statistical function.

For example, item_price, As we can see, data type for item_price is Object, we can convert it into Float by using lambda function as follows :

Lambda Function :

This is a way to create small anonymous functions, i.e. functions without a name. They are just needed where they have been created.
Lets break it down :

        lambda : — this is a lambda function
             x : — the parameter name within the function
float(x[1:-1]) : — what to do with the parameter


The .apply() method is going through every record one-by-one in the df.item_price series, where ip is each record. 
The lambda function converts the value of each df.item_price record in float.

They are especially convenient in data analysis because, as you’ll see, there are many cases where data transformation functions will take functions as arguments. It’s often less typing (and clearer) to pass a lambda function as opposed to writing a full-out function declaration or even assigning the lambda function to a local variable.

This might be a strange pattern to see the first few times, but when you’re writing short functions, the lambda function allows you to work more quickly than the other function definition.

In [9]:
ip = lambda x: float(x[1:-1])
df.item_price = df.item_price.apply(ip)

In [10]:
# Now lets verify the data types again 
df.dtypes
df.head(5)

order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
dtype: object

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  
0                                                NaN        2.39  
1                                       [Clementine]        3.39  
2                                            [Apple]        3.39  
3                                                NaN        2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98

Yes. We have the data types for all columns in the required format. Also we got rid of $ from item_price

--------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------------

In [11]:
# To displays a table of detailed distribution of Statistical information for each of the numerical attributes in our data frame, we use the below function

df.describe()

order_id     quantity   item_price
count  4622.000000  4622.000000  4622.000000
mean    927.254868     1.075725     7.464336
std     528.890796     0.410186     4.245557
min       1.000000     1.000000     1.090000
25%     477.250000     1.000000     3.390000
50%     926.000000     1.000000     8.750000
75%    1393.000000     1.000000     9.250000
max    1834.000000    15.000000    44.250000

This dataset is not having much numerical fields to draw up any conclusion from above Statistical Info.
At a time max 15 quantity of items are ordered. And below function (value_counts) shows the count of unique values of any column also indicates the same :)
Lowest price of any item is `$`1.09, max price is `$`44.25

Interpretation is the process of assigning meaning to the collected information and determining the conclusions, significance, and implications of the findings. So lets understand each of these statistical terms from EDA perspective.

Mean – 
The mean score represents a numerical average for a set of responses.For a data set, the terms arithmetic mean, mathematical expectation, and sometimes average are used synonymously to refer to a central value of a discrete set of numbers. If the data set were based on a series of observations obtained by sampling from a statistical population, the arithmetic mean is termed the sample mean to distinguish it from the population mean.
Here mean item_price is $7.46

Standard deviation – 
The standard deviation represents the distribution of the responses around the mean. It indicates the degree of consistency among the responses. The standard deviation, in conjunction with the mean, provides a better understanding of the data. For example, if the mean is 7.46 with a standard deviation (StD) of 4.24, then two-thirds of the item_price lie between 3.22 (7.46 – 4.24) and 11.7 (7.46 + 4.24).


Other values min, 25`%`, 50`%`, 75`%` & max helps us to gain maximum insight into the data set and its underlying structure.


--------------------------------------------------------------------------------------------------------------------------------



Lets learn few basic statistics functions 

df.describe()  --  Summary statistics for numerical columns

df.mean()      --  Return the mean of all columns

df.corr()      --  Finds the correlation between columns in a DataFrame

df.count()     --  Counts the number of non-null values in each DataFrame column

df.max()       --  Finds the highest value in each column

df.min()       --  Finds the lowest value in each column

df.median()    --  Finds the median of each column

df.std()       --  Finds the standard deviation of each column

In [14]:
# Statistical Information :

print('Display Mean of all available Columns : \n' + str(df.mean()) + '\n')
print('Display  correlation between columns in a DataFrame : \n' + str(df.corr()) + '\n')
print('Display counts of non-null values in each DataFrame column : \n' + str(df.count()) + '\n')
print('Display highest value in each column : \n' + str(df.max()) + '\n')
print('Display lowest value in each column : \n' + str(df.min()) + '\n')
print('Display median of each column : \n' + str(df.median()) + '\n')
print('Display median of each column : \n' + str(df.std()))

Display Mean of all available Columns : 
order_id      927.254868
quantity        1.075725
item_price      7.464336
dtype: float64

Display  correlation between columns in a DataFrame : 
            order_id  quantity  item_price
order_id    1.000000  0.032397   -0.001618
quantity    0.032397  1.000000    0.263981
item_price -0.001618  0.263981    1.000000

Display counts of non-null values in each DataFrame column : 
order_id              4622
quantity              4622
item_name             4622
choice_description    3376
item_price            4622
dtype: int64

Display highest value in each column : 
order_id                                                           1834
quantity                                                             15
item_name                                             Veggie Soft Tacos
choice_description    [[Tomatillo-Red Chili Salsa (Hot), Tomatillo-G...
item_price                                                        44.25
dtype: object

Display lowest

As EDA is used to identify systematic relations between variables when there are no (or not complete) a expectations as to the nature of those relations, exploratory data analysis methods include both simple basic statistics and more advanced techniques designed to identify patterns in multivariate data sets.

This statistical info helps to identify distribution of variables, skew or non-normal patterns in data, reviewing correlation (Correlation is a measure of the relation between two or more variables) for coefficients that meet certain thresholds, or examining data slice by slice systematically reviewing combinations of levels of control variables.

--------------------------------------------------------------------------------------------------------------------------------


As apparent from counts of non-null values in each column there are around (4622 - 3376 =) 1246 nulls present in choice_description column.

Some conclusions might not report any information on outcomes of interest to the review. For example, there may be no information on quality, or on serious adverse effects. It is often difficult to determine whether this is because the outcome was not measured or because the outcome was not reported.

Reasons for handling NaNs :
• One might end up drawing an inaccurate inference about the data
• Due to improper handling, the result obtained will differ from ones where the missing values are present
• Probability cannot be predicted from the variables in the model which involves those types of missing values that are not randomly distributed across the observations

Now we will check various functions to see if there are any null values present in any other columns 

df.isnull()                 -- Returns False for not null values in entire dataset

df['column_name'].notnull() -- Returns True for not null columns else returns False for all rows in the mentioned column



In case if you want to insert some null values in Dataset we have below function :

df.iloc[3:5,0] = np.nan  -- This will insert NaN in 3rd & 4th row's 0th (1st) column
df.iloc[5:8,2] = np.nan  -- This will insert NaN in 5th, 6th & 7th row's 2nd (3rd) column

--------------------------------------------------------------------------------------------------------------------------------

Points to Remember for NaNs :

When summing data, NA (missing) values will be treated as zero
If the data are all NA, the result will be NA
Methods like cumsum and cumprod ignore NA values, but preserve them in the resulting arrays
NA groups in GroupBy are automatically excluded


In [15]:
df['item_name'].notnull().tail(5)

4617    True
4618    True
4619    True
4620    True
4621    True
Name: item_name, dtype: bool

In [16]:
df.isnull().head(5)

order_id quantity item_name choice_description item_price
0    False    False     False               True      False
1    False    False     False              False      False
2    False    False     False              False      False
3    False    False     False               True      False
4    False    False     False              False      False

As a general rule, most methodologists believe that missing data should not be used as a reason to exclude a study from a systematic review. It is more appropriate to include the study in the review, and to discuss the potential implications of its absence from a meta-analysis.


Pandas objects are equipped with various data manipulation methods for dealing with missing data by either filling the missing data with some values ( constant value, mean value, value from previous row etc) or we can drop the missing value

-------------------------------------------------------------------------------------------------------------------------------

To simply exclude labels from a data set which refer to missing data, use the dropna method:

df.dropna(axis=0) -- Drop entire row which contain NaN values
df.dropna(axis=1) -- Drop entire column which contain NaN values

By default, dropna() will drop all rows in which any null value is present

But this drops some good data as well; one might rather be interested in dropping rows or columns with all NA values, or a majority of NA values. This can be specified through the how or thresh parameters, which allow fine control of the number of nulls to allow through.

The default is how='any', such that any row or column (depending on the axis keyword) containing a null value will be dropped. You can also specify how='all', which will only drop rows/columns which are all null values.

df.dropna(axis=1, how='all')

For finer control, the thresh parameter lets us specify a minimum number of non-null values for the row/column to be kept.

df.dropna(thresh=3)

Here the only those row will be dropped which contain only two non-null values.


-------------------------------------------------------------------------------------------------------------------------------

Sometimes rather than dropping NA values, you’d rather replace them with a valid value. This value might be a single number like zero, or it might be some sort of imputation or interpolation from the good values


Filling the Missing Values :
The fillna function can “fill in” NA values with non-null data in a couple of ways & returns a copy of the array with the null values replaced :

df.fillna(0)                        -- Fill all the NaN values with Zero
df['column_name'].fillna('missing') -- Fill all the NaN values from mentioned column_name with "missing"

We can propagate non-null values forward or backward
df.fillna(method='pad', limit=1)    -- pad / ffill      Fill values forward; limit is for limiting the amount of fillings 
                                       bfill / backfill Fill values backward
dff.fillna(dff.mean())              -- Fill NaN from all columns with mean value
dff.fillna(dff.mean()['B':'C'])     -- Fill NaN from all column B & C with mean value

We can specify a forward-fill to propagate the previous value forward : data.fillna(method='ffill')

we can also specify an axis along-which a back-fill to propagate the next values backward : df.fillna(method='bfill', axis=1)

Please Note : 
If a previous/next value is not available during a forward fill / backward fill, the NA value remains.

--------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------------


Coming now are the ways to slice & dice the data.

The axis labeling information in pandas objects serves many purposes:

• Identifies data (i.e. provides metadata) using known indicators, important for analysis, visualization, and interactive console display 
• Enables automatic and explicit data alignment
• Allows intuitive getting and setting of subsets of the data set
• Preliminary selection of appropriate models
• Detection of mistakes
--------------------------------------------------------------------------------------------------------------------------------

In Python, indexing operators [] and attribute operator . provide quick and easy access to pandas data structures across a wide range of use cases. 

Now we will see various ways to select the data. For this we have below functions :

df['Column_Name']  -- Returns column with label Column_Name as Series

df[[col1, col2]] -- Return Columns as a new DataFrame

df.iloc[0]       -- Selection by position (integer based indexing)

df.loc[0]        -- Selection by label

df.iloc[0,:]     -- Returns first row

df.iloc[0,0]     -- Returns first element of first column

--------------------------------------------------------------------------------------------------------------------------------

Few points to remember about Index :

Add a date index :
df.index = pd.date_range('1900/1/30', periods=df.shape[0])

In [13]:
print('Display Column :\n' + str(df['item_name']))
print('Display any combination of Columns : \n' + str(df[['item_name','choice_description']]))
# Display only 2 columns
rev_left = df.loc[:, [ 'item_name', 'item_price' ]]
rev_left.tail(6)

print('Display any row from the Dataset by mentioning the position/row number :\n' + str(df.iloc[34])) 
#this will display 34th row from the Dataset
print('Display any row from the Dataset by mentioning the index :\n' + str(df.loc[34])) 
# Both functions, loc[9] & iloc[9] display same result

Display Column :
0                Chips and Fresh Tomato Salsa
1                                        Izze
2                            Nantucket Nectar
3       Chips and Tomatillo-Green Chili Salsa
4                                Chicken Bowl
5                                Chicken Bowl
6                               Side of Chips
7                               Steak Burrito
8                            Steak Soft Tacos
9                               Steak Burrito
10                        Chips and Guacamole
11                       Chicken Crispy Tacos
12                         Chicken Soft Tacos
13                               Chicken Bowl
14                        Chips and Guacamole
15      Chips and Tomatillo-Green Chili Salsa
16                            Chicken Burrito
17                            Chicken Burrito
18                                Canned Soda
19                               Chicken Bowl
20                        Chips and Guacamole
21               

item_name  item_price
4616  Chips and Guacamole        4.45
4617        Steak Burrito       11.75
4618        Steak Burrito       11.75
4619   Chicken Salad Bowl       11.25
4620   Chicken Salad Bowl        8.75
4621   Chicken Salad Bowl        8.75

Display any row from the Dataset by mentioning the position/row number :
order_id                         17
quantity                          1
item_name             Bottled Water
choice_description              NaN
item_price                     1.09
Name: 34, dtype: object
Display any row from the Dataset by mentioning the index :
order_id                         17
quantity                          1
item_name             Bottled Water
choice_description              NaN
item_price                     1.09
Name: 34, dtype: object


--------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------------


   As we are analyzing this dataset for understanding customer's preference for food items and the quantity of items ordered which will help the Shop owner decide the requirements of basic raw materials they need to keep in stock, we will not skip the NaNs which are around 1/4th of total rows and that too only for choice_description column.
   
   Lets apply various Filter, Sort or Group By functions to analyse data which would help us to figure out the trends for customer's prefrence

In [18]:
print('Most ordered item in choice_description column: \n' + str(df.choice_description.value_counts().head(5)))

Most ordered item in choice_description column: 
[Diet Coke]                                                                          134
[Coke]                                                                               123
[Sprite]                                                                              77
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Lettuce]]                42
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Guacamole, Lettuce]]     40
Name: choice_description, dtype: int64


In [17]:
print('Most ordered item in item_name column: \n' + str(df.item_name.value_counts().head(5)))

Most ordered item in item_name column: 
Chicken Bowl           726
Chicken Burrito        553
Chips and Guacamole    479
Steak Burrito          368
Canned Soft Drink      301
Name: item_name, dtype: int64


In [12]:
--------------------------------------------------------------------------------------------------------------------------------
# value_counts function Returns object containing counts of unique values.
df.quantity.value_counts(dropna=False)

1     4355
2      224
3       28
4       10
15       1
7        1
10       1
5        1
8        1
Name: quantity, dtype: int64

This shows for quantity column we have 10 unique values for which various quantities are ordered. Majority of items are ordered only once. There is just 1 item ordered whose quantity is 15 (max quantity)

This would help the owner know the frequency of the various items ordered.
We will further explore the data before drawing any specific conclusion.

In [19]:
print('Total items ordered :\n' + str(df.quantity.sum()))

Total items ordered :
4972


In [20]:
print('TYPE 1 \n Total number of Orders : ' + str(df.order_id.value_counts().count()) + '\n')

orders = len((set([order for order in df.order_id])))
print('TYPE 2 \n Number of orders made in the period: ' + str(orders))

TYPE 1 
 Total number of Orders : 1834

TYPE 2 
 Number of orders made in the period: 1834


In [21]:
print('Total revenue generated :\n' + str(df.item_price.sum()))

Total revenue generated :
34500.16


In [22]:
print('TYPE 1 \n Number of different items sold: ' + str(len(df.item_name.unique())) + '\n')

print('TYPE 2 \n Number of different items sold: ' + str(df.item_name.value_counts().count()))

TYPE 1 
 Number of different items sold: 50

TYPE 2 
 Number of different items sold: 50


In [ ]:
# make the comparison
chipo10 = df[df['item_price'] > 10.00]
chipo10.head()

len(chipo10)

In [23]:
#AS: Average Cost or Average Price?
avg_order = df.item_price.sum() / df.order_id.value_counts().count()

print "The average cost per order is: $%s" % round(avg_order, 2)

The average cost per order is: $18.81


In [ ]:
#Slicing the dataset Column-wise
# delete the duplicates in item_name and quantity
chipo_filtered = df.drop_duplicates(['item_name','quantity'])

# select only the ones with quantity equals to 1
price_per_item = chipo_filtered[chipo_filtered.quantity == 1]

#
price_per_item = price_per_item[['item_name', 'item_price']]

# sort the values from the most to less expensive
price_per_item.sort_values(by = "item_price", ascending = False)

In [ ]:
print "Quantity of most expensive item ordered : " 
df.sort_values(by = "item_price", ascending = False).head(1)

In [ ]:
print "Number of times people orderd more than one Canned Soda"
# It can be any item
chipo_m1 = df[(df.item_name == "Canned Soda") & (df.quantity > 1)]
len(chipo_m1)

In [ ]:
print "Number of times Veggie Salad Bowl is ordered"
# It can be any item

chipo_salad = df[df.item_name == "Veggie Salad Bowl"]
len(chipo_salad)

  This all analysis is a part of EDA and it helps to know the data by letting us know the customer prefernce which will eventually help the owner to increase the revenue
  
  For example, as people have ordered more than one canned soda 20 times, Owner can decide on how many canned soda they need to keep in stock. 
  
  Also Veggie Salad Bowl is ordered 18 times, owner can think of improving the test by adding some extra sauce /more choices to it.

In [24]:
# As we are aware that NaNs in group by are automatically excluded, we are not explicitly dropping NaN values

item_nm = df.groupby('item_name')

Grouping helps EDA in below ways :

1.Grouping data allows characteristics of the data to be more easily interpreted than would be true if the raw data were to be examined. 
2.Improves the accuracy/efficiency of estimation. 
3.Permits greater balancing of statistical power of tests of differences between strata by sampling equal numbers from strata varying widely in size.

NOTE:
Strata refers to sampling method in which the total population is divided into non-overlapping subgroups.

In [25]:
# To understand the total quantity ordered across each item & the total cost for that item we use below function.
dfn = item_nm.sum()
dfn

order_id  quantity  item_price
item_name                                            
6 Pack Soft Drink         52322        55      356.95
Barbacoa Bowl             53972        66      672.36
Barbacoa Burrito          74718        91      894.75
Barbacoa Crispy Tacos      5613        12      120.21
Barbacoa Salad Bowl        9708        10      106.40
Barbacoa Soft Tacos       18725        25      250.46
Bowl                        472         4       29.60
Burrito                    1550         6       44.40
Canned Soda               76396       126      137.34
Canned Soft Drink        304753       351      438.75
Carnitas Bowl             62742        71      736.71
Carnitas Burrito          51652        60      597.83
Carnitas Crispy Tacos      5868         8       77.96
Carnitas Salad             1500         1        8.99
Carnitas Salad Bowl        6693         6       66.34
Carnitas Soft Tacos       37673        40      375.94
Chicken Bowl             713926       761     7342.73
Chicken Burrito          497303       591     5575.82
Chicken Crispy Tacos      47382        50      472.13
Chicken Salad             10465         9       81.09
Chicken Salad Bowl       117104       123     1228.75
Chicken Soft Tacos        98395       120     1108.09
Crispy Tacos                434         2       14.80
Izze                      10587        20       67.80
Nantucket Nectar          21814        29       98.31
Salad                      1150         2       14.80
Steak Bowl               193752       221     2260.19
Steak Burrito            328437       386     3851.43
Steak Crispy Tacos        35818        36      357.34
Steak Salad                2362         4       35.66
Steak Salad Bowl          29590        31      343.59
Steak Soft Tacos          50177        56      536.05
Veggie Bowl               72270        87      867.99
Veggie Burrito            80962        97      934.77
Veggie Crispy Tacos         668         1        8.49
Veggie Salad               7265         6       50.94
Veggie Salad Bowl         17303        18      182.50
Veggie Soft Tacos          5520         8       73.96

In [26]:
# Sorting the above list based on quantity, helps to know the highest/lowest preferred item along with its price
sort_list = dfn.sort_values(['quantity'], ascending=False)
sort_list.head(1)

order_id  quantity  item_price
item_name                                   
Chicken Bowl    713926       761     7342.73

From we this we can help the owner decide the name of those items which they need to stock up as per the quantity based on customer's choice: 

Coke, Chicken, Fresh Tomato Salsa, Cheese, Black Beans are the top most choice so these need to be stored in higher quantity as compared to other items.

In [27]:
print('Most ordered item :' + str(df.item_name.value_counts().head(1)))
print('Count of Most ordered item :' + str(df.item_name.value_counts().max()))

Most ordered item :Chicken Bowl    726
Name: item_name, dtype: int64
Count of Most ordered item :726


In [28]:
cnt = df['item_name'].unique()
count = len(cnt)
print('Count of items ordered: ' + str(count))

Count of items ordered: 50


Lets group by choice_description column and then display the top 5 items. It is another way of slicing the data from another perspective (choice_description) to know more about the data

We are doing all this as a part of EDA to get insights of our data which would help us to draw some conclusion about customer prefernce and accordingly we can advise the owner to help them grow their revenue.

Now for grouping data based on choice_description.

In [29]:
# Get rid of NaNs
# NaNs in group by are automatically excluded. This step can be excluded.
df2 = df.dropna(how='any')
print('After dropping NaNs : ' )
df2.head(2)

# Group Data by choice_description
ch_des = df2.groupby('choice_description')
# Do summation across groups. Again to know the most preferred choce and the cost. 
dfn1 = ch_des.sum()
print('\nAfter grouping Data across choice_description : ')
dfn1.head(2)

# Sort data based on the quantity ordered to display top 'N' & bottom 'N' records
sort_list_cd = dfn1.sort_values(['quantity'], ascending=False)
print('\n Top 5 items after sorting the grouped Data by quantity : ')
sort_list_cd.head(50)
print('\n Last 5 items after sorting the grouped Data by quantity : ')
sort_list_cd.tail(50)

# Sort data based on the quantity ordered to display top 'N' & bottom 'N' records
sort_list_ip = dfn1.sort_values(['item_price'], ascending=False)
print('\n Top 5 items after sorting the grouped Data by item_price : ')
sort_list_ip.head(25)
print('\n Last 5 items after sorting the grouped Data by item_price : ')
sort_list_ip.tail(25)

After dropping NaNs : 


order_id  quantity         item_name choice_description  item_price
1         1         1              Izze       [Clementine]        3.39
2         1         1  Nantucket Nectar            [Apple]        3.39


After grouping Data across choice_description : 


order_id  quantity  \
choice_description                                                       
[Adobo-Marinated and Grilled Chicken, Pinto Bea...       214         1   
[Adobo-Marinated and Grilled Chicken, [Sour Cre...       279         1   

                                                    item_price  
choice_description                                              
[Adobo-Marinated and Grilled Chicken, Pinto Bea...         7.4  
[Adobo-Marinated and Grilled Chicken, [Sour Cre...         7.4


 Top 5 items after sorting the grouped Data by quantity : 


order_id  quantity  \
choice_description                                                       
[Diet Coke]                                           123455       159   
[Coke]                                                122752       143   
[Sprite]                                               80426        89   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     43088        49   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     36041        42   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     37550        40   
[Lemonade]                                             31892        36   
[Fresh Tomato Salsa (Mild), [Pinto Beans, Rice,...     24432        36   
[Coca Cola]                                            19282        32   
[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream,...     29614        30   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese]]      27583        30   
[Fresh Tomato Salsa, [Rice, Cheese, Lettuce]]          23326        26   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     24637        25   
[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]       22664        24   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     17529        23   
[Fresh Tomato Salsa, Rice]                             17175        22   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice]]        21848        22   
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...     20902        21   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     23031        21   
[Dr. Pepper]                                           13223        21   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     21722        21   
[Tomatillo Red Chili Salsa, [Rice, Cheese, Lett...     15512        20   
[Nestea]                                               23349        20   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     24051        20   
[Tomatillo Red Chili Salsa, [Rice, Cheese, Sour...     17854        20   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     17989        19   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...     11846        19   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     22117        19   
[Tomatillo Red Chili Salsa, [Rice, Black Beans,...     15244        18   
[Tomatillo Red Chili Salsa, [Rice, Black Beans,...     13725        18   
[Mountain Dew]                                          8687        18   
[Fresh Tomato Salsa, [Cheese, Lettuce]]                20662        18   
[Tomatillo Red Chili Salsa, [Rice, Black Beans,...     12777        17   
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...     14823        15   
[Fresh Tomato Salsa (Mild), [Rice, Cheese, Sour...     11646        15   
[Fresh Tomato Salsa, [Rice, Pinto Beans, Cheese]]      12214        14   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...      8794        14   
[Fresh Tomato Salsa (Mild), [Black Beans, Rice,...      9576        14   
[Roasted Chili Corn Salsa (Medium), [Rice, Faji...     13335        14   
[Diet Dr. Pepper]                                      10405        14   
[Fresh Tomato Salsa]                                   11926        13   
[Tomatillo-Red Chili Salsa (Hot), [Black Beans,...      6501        13   
[Fresh Tomato Salsa, [Rice, Pinto Beans, Cheese...     11235        12   
[Tomatillo-Red Chili Salsa (Hot), [Rice, Cheese...      7931        12   
[Roasted Chili Corn Salsa, [Rice, Cheese, Sour ...     11747        12   
[Tomatillo-Red Chili Salsa (Hot), [Black Beans,...     12498        12   
[Tomatillo Green Chili Salsa, [Rice, Pinto Bean...     12546        12   
[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream,...     14903        12   
[Tomatillo Red Chili Salsa, [Rice, Black Beans,...      8268        11   
[Tomatillo Red Chili Salsa, [Fajita Vegetables,...     11285        11   

                                                    item_price  
choice_description                                              
[Diet Coke]                                             326.71  
[Coke]  


 Last 5 items after sorting the grouped Data by quantity : 


order_id  quantity  \
choice_description                                                       
[Roasted Chili Corn Salsa (Medium), [Rice, Faji...      1288         1   
[Roasted Chili Corn Salsa (Medium), [Rice, Guac...       900         1   
[Roasted Chili Corn Salsa (Medium), [Rice, Pint...      1277         1   
[Roasted Chili Corn Salsa (Medium), [Rice, Pint...      1452         1   
[Roasted Chili Corn Salsa (Medium), [Rice, Sour...       356         1   
[Roasted Chili Corn Salsa, Cheese]                      1152         1   
[Roasted Chili Corn Salsa, Rice]                          18         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1714         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1662         1   
[Roasted Chili Corn Salsa, [Rice, Fajita Vegeta...       387         1   
[Roasted Chili Corn Salsa, [Lettuce, Sour Cream...      1011         1   
[Roasted Chili Corn Salsa, [Pinto Beans, Cheese...      1816         1   
[Roasted Chili Corn Salsa, [Pinto Beans, Rice, ...      1150         1   
[Roasted Chili Corn Salsa, [Pinto Beans, Rice, ...      1553         1   
[Roasted Chili Corn Salsa, [Pinto Beans, Sour C...      1825         1   
[Adobo-Marinated and Grilled Chicken, [Sour Cre...       279         1   
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...      1625         1   
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...      1301         1   
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...       211         1   
[Roasted Chili Corn Salsa, [Rice, Cheese, Guaca...       855         1   
[Roasted Chili Corn Salsa, [Rice, Cheese, Sour ...      1018         1   
[Roasted Chili Corn Salsa, [Rice, Fajita Vegeta...       192         1   
[Roasted Chili Corn Salsa, [Rice, Fajita Vegeta...        65         1   
[Roasted Chili Corn Salsa, [Rice, Fajita Vegeta...       387         1   
[Roasted Chili Corn Salsa, [Rice, Fajita Vegeta...      1761         1   
[Roasted Chili Corn Salsa, [Rice, Fajita Vegeta...      1134         1   
[Roasted Chili Corn Salsa, [Rice, Fajita Vegeta...      1269         1   
[Roasted Chili Corn Salsa, [Rice, Fajita Vegeta...       235         1   
[Roasted Chili Corn Salsa, [Pinto Beans, Cheese...      1483         1   
[Roasted Chili Corn Salsa, [Lettuce, Cheese]]            853         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...       362         1   
[Roasted Chili Corn Salsa, [Lettuce, Black Bean...       715         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...       621         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1393         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1016         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1723         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1679         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1403         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1659         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1138         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1203         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...       125         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...       691         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1176         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      1762         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...       577         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...       585         1   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...       235         1   
[Roasted Chili Corn Salsa, [Guacamole, Sour Cre...       987         1   
[[Tomatillo-Red Chili Salsa (Hot), Tomatillo-Gr...      1299         1   

                                                    item_price  
choice_description                                              
[Roasted Chili Corn Salsa (Medium), [Rice, Faji...        8.49  
[Roasted


 Top 5 items after sorting the grouped Data by item_price : 


order_id  quantity  \
choice_description                                                       
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     37550        40   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     43088        49   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     36041        42   
[Diet Coke]                                           123455       159   
[Fresh Tomato Salsa (Mild), [Pinto Beans, Rice,...     24432        36   
[Coke]                                                122752       143   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese]]      27583        30   
[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream,...     29614        30   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     17529        23   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     23031        21   
[Fresh Tomato Salsa, [Rice, Cheese, Lettuce]]          23326        26   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     24051        20   
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     24637        25   
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...     11846        19   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     17989        19   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     22117        19   
[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]       22664        24   
[Tomatillo Red Chili Salsa, [Rice, Black Beans,...     12777        17   
[Fresh Tomato Salsa, Rice]                             17175        22   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice]]        21848        22   
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...     20902        21   
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     21722        21   
[Tomatillo Red Chili Salsa, [Rice, Cheese, Sour...     17854        20   
[Tomatillo Red Chili Salsa, [Rice, Cheese, Lett...     15512        20   
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...     14823        15   

                                                    item_price  
choice_description                                              
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      455.50  
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      432.25  
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      372.64  
[Diet Coke]                                             326.71  
[Fresh Tomato Salsa (Mild), [Pinto Beans, Rice,...      311.64  
[Coke]                                                  288.79  
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese]]       267.00  
[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream,...      265.00  
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      264.25  
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      238.25  
[Fresh Tomato Salsa, [Rice, Cheese, Lettuce]]           229.50  
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...      227.00  
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      222.75  
[Roasted Chili Corn Salsa, [Fajita Vegetables, ...      220.03  
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...      216.89  
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...      215.39  
[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]        213.14  
[Tomatillo Red Chili Salsa, [Rice, Black Beans,...      194.89  
[Fresh Tomato Salsa, Rice]                              194.00  
[Fresh Tomato Salsa, [Fajita Vegetables, Rice]]         192.50  
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...      187.25  
[Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...      186.25  
[Tomatillo Red Chili Salsa, [Rice, Cheese, Sour...      183.00  
[Tomatillo Red Chili Salsa, [Rice, Cheese, Lett...      182.00  
[Roasted Chili Corn Salsa, [Rice, Black Beans, ...      171.25


 Last 5 items after sorting the grouped Data by item_price : 


order_id  quantity  \
choice_description                                                       
[[Fresh Tomato Salsa (Mild), Roasted Chili Corn...      1099         1   
[[Fresh Tomato Salsa (Mild), Roasted Chili Corn...       594         1   
[[Fresh Tomato Salsa (Mild), Roasted Chili Corn...      1363         1   
[[Fresh Tomato Salsa (Mild), Roasted Chili Corn...       742         1   
[Roasted Chili Corn (Medium), [Lettuce, Black B...      1598         1   
[Roasted Chili Corn (Medium), [Lettuce, Black B...      1598         1   
[Rice]                                                  1533         1   
[Fresh Tomato (Mild), [Lettuce, Fajita Veggies,...      1466         1   
[Fresh Tomato (Mild), [Lettuce, Fajita Veggies,...      1237         1   
[Fresh Tomato (Mild), [Lettuce, Rice, Fajita Ve...       479         1   
[Fresh Tomato (Mild), [Rice, Sour Cream, Cheese]]       1160         1   
[[Lettuce, Rice, Fajita Veggies]]                       1628         1   
[Tomatillo Green Chili (Medium), [Lettuce, Blac...      1598         1   
[Tomatillo Red Chili (Hot), [Lettuce, Rice, Che...       517         1   
[White Rice, Adobo-Marinated and Grilled Chicke...       575         1   
[Adobo-Marinated and Grilled Chicken, [Sour Cre...       279         1   
[White Rice, Adobo-Marinated and Grilled Steak,...       313         1   
[White Rice, Adobo-Marinated and Grilled Steak,...       313         1   
[Brown Rice, Adobo-Marinated and Grilled Chicke...       575         1   
[Braised Barbacoa, Vegetarian Black Beans, [Sou...       214         1   
[Braised Barbacoa, Pinto Beans, [Sour Cream, Sa...       217         1   
[Adobo-Marinated and Grilled Steak]                      217         1   
[Adobo-Marinated and Grilled Steak, [Sour Cream...       279         1   
[Adobo-Marinated and Grilled Chicken]                    217         1   
[Adobo-Marinated and Grilled Chicken, Pinto Bea...       214         1   

                                                    item_price  
choice_description                                              
[[Fresh Tomato Salsa (Mild), Roasted Chili Corn...        8.49  
[[Fresh Tomato Salsa (Mild), Roasted Chili Corn...        8.49  
[[Fresh Tomato Salsa (Mild), Roasted Chili Corn...        8.49  
[[Fresh Tomato Salsa (Mild), Roasted Chili Corn...        8.49  
[Roasted Chili Corn (Medium), [Lettuce, Black B...        8.19  
[Roasted Chili Corn (Medium), [Lettuce, Black B...        8.19  
[Rice]                                                    8.19  
[Fresh Tomato (Mild), [Lettuce, Fajita Veggies,...        8.19  
[Fresh Tomato (Mild), [Lettuce, Fajita Veggies,...        8.19  
[Fresh Tomato (Mild), [Lettuce, Rice, Fajita Ve...        8.19  
[Fresh Tomato (Mild), [Rice, Sour Cream, Cheese]]         8.19  
[[Lettuce, Rice, Fajita Veggies]]                         8.19  
[Tomatillo Green Chili (Medium), [Lettuce, Blac...        8.19  
[Tomatillo Red Chili (Hot), [Lettuce, Rice, Che...        8.19  
[White Rice, Adobo-Marinated and Grilled Chicke...        7.40  
[Adobo-Marinated and Grilled Chicken, [Sour Cre...        7.40  
[White Rice, Adobo-Marinated and Grilled Steak,...        7.40  
[White Rice, Adobo-Marinated and Grilled Steak,...        7.40  
[Brown Rice, Adobo-Marinated and Grilled Chicke...        7.40  
[Braised Barbacoa, Vegetarian Black Beans, [Sou...        7.40  
[Braised Barbacoa, Pinto Beans, [Sour Cream, Sa...        7.40  
[Adobo-Marinated and Grilled Steak]                       7.40  
[Adobo-Marinated and Grilled Steak, [Sour Cream...        7.40  
[Adobo-Marinated and Grilled Chicken]                     7.40  
[Adobo-Marinated and Grilled Chicken, Pinto Bea...        7.40

Lets explore some more data by grouping on both item_name & choice_description

In [30]:
# Grouping by 2 columns would help us to know the most preferred combination
item_tot = df.groupby(['item_name', 'choice_description']).sum()
it3 = item_tot.dropna(how='any')

#Sorting the grouped data based on item_price
sort_list_ic = it3.sort_values(['item_price'], ascending=False)
print('\n Top 10 items after sorting the grouped Data by item_price :')
sort_list_ic.head(100)
print('\n Last 10 items after sorting the grouped Data by item_price :')
sort_list_ic.tail(100)


 Top 10 items after sorting the grouped Data by item_price :


order_id  \
item_name          choice_description                                             
Chicken Burrito    [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     17682   
                   [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     13781   
6 Pack Soft Drink  [Diet Coke]                                            24539   
Chicken Bowl       [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     15003   
Canned Soft Drink  [Coke]                                                103312   
Chicken Bowl       [Fresh Tomato Salsa, [Fajita Vegetables, Rice]]        14095   
Canned Soft Drink  [Diet Coke]                                            88465   
Chicken Bowl       [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     10990   
6 Pack Soft Drink  [Coke]                                                 19440   
Chicken Burrito    [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      8226   
                   [Fresh Tomato Salsa, Rice]                              8653   
Chicken Bowl       [Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...      8794   
                   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream,...     11562   
                   [Fresh Tomato Salsa, [Rice, Cheese, Lettuce]]          11593   
                   [Fresh Tomato Salsa, [Rice, Black Beans, Cheese]]      11661   
Chicken Salad Bowl [Tomatillo Green Chili Salsa, [Black Beans, Che...      3744   
Steak Burrito      [Fresh Tomato Salsa (Mild), [Pinto Beans, Rice,...     10833   
Chicken Bowl       [Tomatillo Green Chili Salsa, [Rice, Pinto Bean...     12546   
                   [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...     11621   
Veggie Burrito     [Fresh Tomato Salsa (Mild), [Pinto Beans, Rice,...      4665   
Chicken Bowl       [Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...     10684   
                   [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      7595   
                   [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      9947   
                   [Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...      7476   
                   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]        9177   
                   [Roasted Chili Corn Salsa, [Rice, Black Beans, ...     10664   
Chicken Burrito    [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream,...     12486   
                   [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      8204   
                   [Tomatillo Red Chili Salsa, [Rice, Black Beans,...      5889   
Chicken Bowl       [[Roasted Chili Corn Salsa (Medium), Fresh Toma...      6408   
...                                                                         ...   
Veggie Burrito     [Fresh Tomato Salsa, [Rice, Pinto Beans, Cheese]]       5959   
Steak Burrito      [Fresh Tomato Salsa, [Rice, Pinto Beans, Cheese...      4638   
                   [Roasted Chili Corn Salsa, [Rice, Pinto Beans, ...      2747   
                   [Tomatillo Green Chili Salsa, [Fajita Vegetable...      2385   
Barbacoa Bowl      [Tomatillo Red Chili Salsa, [Rice, Cheese, Lett...      2900   
Steak Soft Tacos   [Fresh Tomato Salsa, [Cheese, Sour Cream]]              1907   
Chicken Bowl       [Fresh Tomato Salsa, [Rice, Fajita Vegetables, ...      2824   
Chicken Burrito    [Fresh Tomato Salsa, [Fajita Vegetables, Rice, ...      6564   
Chicken Bowl       [Roasted Chili Corn Salsa, [Rice, Black Beans, ...      3265   
                   [Tomatillo Green Chili Salsa, [Rice, Sour Cream...      1383   
Steak Bowl         [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sour...      5867   
Steak Burrito      [Tomatillo-Red Chili Salsa (Hot), [Rice, Cheese...      2941   
                   [[Roasted Chili Corn Salsa (Medium), Tomatillo-...      5896   
                   [Roasted Chili Corn Salsa (Medium), [Rice, Blac...       649   
Chicken Burrito    [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...      3604   
Chicken Bowl       [Fresh Tomato Salsa, [Rice, Black Beans, Cheese...      4854   
Chic


 Last 10 items after sorting the grouped Data by item_price :


order_id  \
item_name          choice_description                                             
Chicken Bowl       [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...       988   
                   [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...       274   
Chicken Burrito    [Tomatillo-Red Chili Salsa (Hot), [Pinto Beans,...       908   
                   [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...      1491   
                   [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...      1555   
                   [Tomatillo-Green Chili Salsa (Medium), [Cheese,...       959   
Chicken Bowl       [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sour...       246   
                   [Fresh Tomato Salsa (Mild), [Rice, Fajita Veggi...      1039   
                   [Fresh Tomato Salsa (Mild), [Rice, Fajita Veggi...       131   
                   [Fresh Tomato Salsa (Mild), [Black Beans, Rice,...       623   
                   [Fresh Tomato Salsa (Mild), Lettuce]                    1271   
                   [Fresh Tomato Salsa (Mild), Cheese]                     1271   
Chicken Burrito    [Tomatillo-Green Chili Salsa (Medium), Lettuce]           78   
                   [Tomatillo-Green Chili Salsa (Medium), [Black B...      1065   
                   [Tomatillo-Green Chili Salsa (Medium), [Black B...       868   
                   [Tomatillo-Green Chili Salsa (Medium), [Black B...      1440   
                   [Tomatillo-Green Chili Salsa (Medium), [Black B...       344   
                   [Tomatillo-Green Chili Salsa (Medium), [Pinto B...         8   
                   [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...      1536   
Chicken Bowl       [Roasted Chili Corn Salsa (Medium), Cheese]             1271   
Chicken Burrito    [Tomatillo-Green Chili Salsa (Medium), [Pinto B...       595   
                   [Tomatillo-Green Chili Salsa (Medium), [Rice, F...        81   
Chicken Bowl       [Roasted Chili Corn Salsa (Medium), [Pinto Bean...      1528   
                   [Roasted Chili Corn Salsa (Medium), [Pinto Bean...       983   
                   [Roasted Chili Corn Salsa (Medium), [Pinto Bean...        13   
Chicken Burrito    [Tomatillo-Green Chili Salsa (Medium), [Rice, F...      1060   
Chicken Bowl       [Roasted Chili Corn Salsa (Medium), [Rice, Faji...       703   
                   [Roasted Chili Corn Salsa (Medium), [Rice, Faji...      1288   
Chicken Burrito    [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...      1095   
                   [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...       477   
...                                                                         ...   
Chicken Soft Tacos [Tomatillo-Green Chili Salsa (Medium), [Pinto B...       287   
                   [Tomatillo-Green Chili Salsa (Medium), [Cheese,...      1433   
                   [Tomatillo-Red Chili Salsa (Hot), [Black Beans,...       555   
Chicken Bowl       [Tomatillo-Red Chili Salsa (Hot), [Rice, Fajita...       651   
                   [[Fresh Tomato Salsa (Mild), Roasted Chili Corn...       689   
                   [[Fresh Tomato Salsa (Mild), Roasted Chili Corn...       623   
                   [[Fresh Tomato Salsa (Mild), Roasted Chili Corn...       779   
                   [Fresh Tomato (Mild), [Lettuce, Fajita Veggies,...      1466   
Chicken Burrito    [Rice]                                                  1533   
                   [Fresh Tomato (Mild), [Lettuce, Rice, Fajita Ve...       479   
                   [Tomatillo Red Chili (Hot), [Lettuce, Rice, Che...       517   
                   [Fresh Tomato (Mild), [Black Beans, Rice, Sour ...      1564   
                   [Tomatillo Green Chili (Medium), [Lettuce, Blac...      1598   
                   [Roasted Chili Corn (Medium), [Lettuce, Black B...      1598   
Chicken Bowl       [Fresh Tomato (Mild), [Rice, Sour Cream, Cheese]]       1160   
                   [Fresh Tomato (Mild), [Lettuce, Fajita Veggies,...      1237   
Chic

From this we can make out 


1. Any changes they can make to improve the taste/quality for the bottom most ordered items. Owner can experiment few of our suggestions as follows :
   For example for almost all top 5 item the preferred Salsa is Fresh Tomato Salsa, so if they can add this instead of just   plain Salsa or Roasted Chili Corn Salsa, there is possibility for these bottom most items to go up the list or atleat to generate some more revenue
   For Burrito, there needs to be option for Black Beans & Fresh Tomato Salsa
   Roasted Chili Corn Salsa seems to be least fav so if possible can be replaced by other fav Salsa (e.g. Fresh Tomato Salsa) as per cutomer's choice
   Provide more choices for Crispy Tacos
   
2. Reduce the stocking quantity of raw materials which are less frequently used.
   
2. Along with these changes owner can reduce the price a bit for few items and then can rise it again based on customer's feedback


All this will help the shop owner to increasse their revenue.

I hope this notebook will help you maximize your efficiency when starting with basic data exploratory analysis in Python.

  Python is really a great tool, and is becoming an increasingly popular language as it is the easiest-to-use languages for rapidly whipping your data into shape. What people are increasingly finding is that Python is a suitable language not only for doingresearch and prototyping but also building the production systems, too.
  
  So, learn Python to perform the full life-cycle of any data science project. It includes reading, analyzing, visualizing and finally making predictions.

  Next notebook we will learn about making plots and static or interactive visualizations which is one of the most important tasks in data analysis
  
If you come across any difficulty while practicing this, or you have any thoughts / suggestions / feedback on the post, please feel free to post them through comments below.